design/one-pager-generic-datasource.md

```diff
@@ -0,0 +1,171 @@
+# Patch from Generic Datasources
+
+* Owner: Maximilian Blatt (maximilian.blatt@accenture.com, @mistermx)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## 1 Background
+
+Crossplane currently supports patching from composite resources to composed
+resources and vice versa. It is also possible to copy values between two
+composed resources inside one compositiion by patching the value from one
+composed back to the composite and then to the second composed resource.
+
+However, Crossplane currently does not provide a way to patch values from
+dynamic sources. Compositions are pratically static right now.
+
+## 2 Current workarounds
+
+While compositions itself are not dynamic, this restrictions can be circumvented
+using one of the following methods:
+
+* `provider-helm`: using [`spec.forProvider.valuesFrom.secretKeyRef`](https://doc.crds.dev/github.com/crossplane-contrib/provider-helm/helm.crossplane.io/Release/v1beta1@v0.10.0#spec-forProvider-valuesFrom-secretKeyRef)
+* `provider-kubernetes`: using [`spec.forProvider.references[].patchesFrom`](https://doc.crds.dev/github.com/crossplane-contrib/provider-kubernetes/kubernetes.crossplane.io/Object/v1alpha1@v0.3.0#spec-references-patchesFrom)
+
+However, none of these methods are Crossplane native and are not available to
+every provider. Furthermore, they would require an additional MR (here `Object`
+or `Release`) to deploy the composed resource increasing the complexity of the
+composition.
+
+## 3 Goals
+
+The goal of this document is to define a proposal for extracting environment
+specific data considering the following requirements:
+
+1. Allow the creation of generic compositions that are able to load environment 
+specific parameters during compose time.
+2. Allow users to define these values on cluster level.
+3. Prevent users from escalting their rights on the cluster and read out fields
+from objects they shouldn't have access to.
+4. (Bonus) Allow package writers to include default values in configuration
+packages.
+
+## 4 Proposed Solution
+
+Introduce a new Crossplane CRD `ValueSet` that can store arbitrary data similar
+to K8s `ConfigMaps` but supports more complex data types:
+
+```yaml
+---
+apiVersion: apiextensions.crossplane.io/v1alpha1
+kind: ValueSet
+metadata:
+  name: sample-value-set
+data:
```

Comment:
I think the proposed structure may not be flexible enough to address with field path. For example, will they use `data.complex.a`? What if there are deeply nested, i.e. the case where you have an array in a `complex` map?

One idea you may want to consider is to use `map[string]runtime.RawExtension` type where you can have an arbitrary YAML object in `RawExtension`. We [use](https://github.com/crossplane/crossplane/blob/79631b8/apis/apiextensions/v1/composition_types.go#L104) that for `resources[*].base` field in `Composition` type. This way anything that can be represented by YAML can be put there and would be addressable via field path (I haven't tried this but fairly sure).

Reply:
I am not sure how flexible `RawExtension` actually is because the data gets parsed into a `runtime.Object`. I am not sure if you can store primitive values in there.

In either case, the [example implementation[(https://github.com/crossplane/crossplane/pull/3007/files#diff-755442f10dbfe9f790b1843c9f660718dcb1bff01a151c41d5ecc74f33cdf6dbR23) mentioned uses the `JSON` struct which allows storing everything that is valid JSON, including complex paths like `data.complex.a`.